Remove functionality

mcdonnelldean opened the pull request:
@mcollina Can you review please? Particularly for any side effects to the filters?

I also went back to the original test style since it matches with all of your other repositories.

mcollina commented on bloomrun.js:
```diff
@@ -43,6 +43,40 @@ BloomRun.prototype.add = function (pattern, payload) {
+  var bucket
+
+  if (matches.length > 0) {
+    bucket = matches[0]
```
You need to iterate here, if the first bucket do not contain the pattern.

mcollina commented on bloomrun.js:
```diff
@@ -43,6 +43,40 @@ BloomRun.prototype.add = function (pattern, payload) {
+    for (var i = 0; i < bucket.data.length; i++) {
+      if (pattern === bucket.data[i].pattern) {
+        bucket.data.splice(i, 1)
+      }
```
You should also check about the second argument off `add`.

mcdonnelldean replied:
Do you mean the payload? It shouldn't matter to check it, it will be copyed if there. Or am I misunderstanding?

mcollina replied:
Given the current API, you can add two elements with the same pattern, but different payload. `remove(payload)` should really be `remove(pattern, payload)`.

mcdonnelldean replied:
Ah! Good catch, all add relevant tests too 😊

mcollina commented on bloomrun.js:
```diff
@@ -43,6 +43,40 @@ BloomRun.prototype.add = function (pattern, payload) {
+      if (bucket === buckets[b]) {
+        this._buckets.splice(b, 1)
+      }
+    }
```
if the bucket contains more than this element, this operation is wrong.

mcollina commented:
Apart from the two little notes, it's all good.

mcdonnelldean commented:
@mcollina super I'll get these worked out tonight and re commit

mcdonnelldean commented:
@mcollina Ok I've updated, I had to break out some functions, could you let me know if you want them anywhere else? For now I just popped them at the top of the file. I added extra tests too.

mcollina commented on README.md:
```diff
@@ -68,6 +69,15 @@ retrieval of objects. If no payload is provided the pattern itself will be
 -------------------------------------------------------
+
+<a name="remove"></a>
+### instance.remove(pattern)
```
Doc is missing the payload here

mcollina commented on bloomrun.js:
```diff
@@ -43,6 +68,33 @@ BloomRun.prototype.add = function (pattern, payload) {
+
+        Object.keys(pattern).forEach(function (key) {
+          that._properties.delete(key)
+        })
```
You can avoid setting `that` and creating a function here, declare them at top scope and pass a `this` parameter to `forEach`. See https://github.com/mcollina/bloomrun/blob/master/lib/genKeys.js for an example (and the docs!)

Allocating functions within loop is particularly bad!
